HOLY FUCKING SHIT, I'VE DONE IT! Map rotation!

MrsTonedOne opened the pull request:
This requires the use of the tgstation-server toolset

The current map is listed in the status and lobby stat panels (regardless of if map rotation is enabled)

Once the shuttle hits the point of no return, the ticker will roll for a chance to do a map rotation.

Rotate chance is calculated as ((round time in minutes)*0.75)%. So a 40minute round has a 30% chance of changing maps. a 60 minute round has a 45% chance.

The percentage part of that equation is configurable 

This is a **random** map rotation that is weighted based on player preferences. Server operators can also rig the weighting to favor certain maps.

Maps can have a min and max population.

A single configured default map will absorb the votes of people who haven't set a preferred map.

Renamed tgstation2 to Box Station

Adds two +SERVER admin verbs:
Force map rotation
Change Map
They do exactly what they say they do.

Adds a next map status tab entry.

MrsTonedOne committed:
map rotation base code

MrsTonedOne committed:
Cleans up map rotation code, removes debugging code

GunHog commented:
Lovely! :+1:

Anonus commented:
VIVA MAP ROTATION

Aloraydrel commented:
aaaaaaaaaah

Alek2ander commented:
I hope this won't cause clients to redownload resources every map change.

Wyzack commented:
Based MSO holy shit amazing 10/10

MrStonedOne commented:
>Does this require the map to be merged for it to be selectable? It appears to, at least.

Yes, sadly. but no reason we can't merge some of these test maps in now.

Cheridan commented:
:rotating_light: :rotating_light:

Jordie0608 commented on code/controllers/configuration.dm:
```diff
@@ -529,6 +535,58 @@
 
+
+/datum/configuration/proc/loadmaplist(filename)
+	var/list/Lines = file2list(filename)
```
You should use bygex for this. :shipit:

MrStonedOne replied:
one, too much work
two, too much work
three, with the amount of conditions i wanted to have, regex wouldn't have worked.

optimumtact replied:
:-1: don't be lazy

hornygranny commented:
sign me the FUCK up  good shit go౦ԁ sHit thats ✔ some goodshit rightth  ere right✔there ✔✔if i do ƽaү so my selｆ 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯  НO0ОଠＯOOＯOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ   💯     Good shit

Supermichael777 commented:
Mr Stoned one for pres-wait hes already our dictator.

Dorsisdwarf commented:
@MrStonedOne 
>I hope this won't cause clients to redownload resources every map change.

Metacide commented:
It's actually happening? My god. Thank you!

MrStonedOne commented:
>I hope this won't cause clients to redownload resources every map change.
in theory, nothing should. Byond is suppose to be able to just download what's changed. Even if there is a difference of files or what not, it should be able to just grab the new or changed files.

In practice, any resource change that happens while another dream seeker window is open has a chance of doing that.

I've narrowed it down to either running ss13 that requires a partial resource download while downloading resources in another ss13 window (say, opening basil and sybil really quickly after an update)(100% chance). Or running ss13 while you have an old idle or disconnected ss13 window open where it had to download resources in the beginning (30% chance).

Cookiesndip commented:
Question, if it rolls for a map rotation, can you 'rotate' to the current map? If not, that's shit.

MrStonedOne commented:
Yes, the script has no care of what the current map is

joonoro commented:
Nice! :+1:

AlexanderUlanH commented:
As a powergaming box fanboy who has to know where all the maintenance loot spawners are on a map to play it this PR troubles me deeply.
:+1:

daman997 commented:
y e s :+1:

Megatrononline commented:
What a horrible idea. All non-box maps give me claustrophobia. Not to mention that you need to get used to a map to play properly so a random map that you are not familiar with means it's just a ruined round for you. I suppose this is good for oldfags, but i play only for two years and what is this.

Anonus commented:
If you're complaining a map rotation feature rotates the map, shoo.

Megatrononline commented:
I mean if it will actually be used

daman997 commented:
>non-box maps give me claustrophodia
box is cramped as fuck compared to meta wat

Incoming5643 commented:
A good idea, but I still think we should just have a server that runs box 24/7 to placate those that enjoy the strong familiarity of it.

Megatrononline commented:
Maybe something to do with a more horizontal layout as oppose to meta's more vertical layout. Or maybe something to do with that don't know it well enough to properly navigate. On box i feel freedom.

Wyzack commented:
>Muh powergayman is slightly lessened on maps i do not know inside and out

welcome to the future nerds, we will drag you kicking and screaming if we have to

Dorsisdwarf commented:
:+1: 
People should get out of the metahabits a server develops from only one map, but few people want to play other stations all the time.

ike709 commented:
:+1: maybe Basil will actually get populated if people can switch it to that ugly box station from time to time.

Dorsisdwarf commented:
I think this PR deprecates basil actually

optimumtact commented:
basil is dead because we no longer have the community size to support it, along with I assume improvements that made playing on a higher pop server less horrible. So the only way to fix it is to attract enough new players to make it worthwhile, perhaps the next summer rush.

Shadowlight213 commented:
Are admins able to force set a new map?

pudl commented:
:-1: There's no purpose.

Iamgoofball commented:
@pudl yes there is and you know it

what is "every single other multiplayer game in existence"

pudl commented:
SS13 isn't a normal multiplayer game. Unless you're looking at absolutely batshit tactical precision, SS13 is a game in which the map you're playing on affects the game more than any other single thing.

Iamgoofball commented:
Every other game has a different play style and system depending on the map. 

Name a few multiplayer games where the map does NOT change how you play the game in the setting.

pudl commented:
Garry's Mod, Minecraft, Euro Truck Simulator 2, Borderlands, every fighting game ever.

Iamgoofball commented:
>garry's mod

gmod isn't a game, it's a platform
even if you are to argue the sandbox bit

your building space/what you build is heavily decided by your map you are on

>Minecraft

you serious right now? the world you spawn in completely determines what you do and what resources you have

>Euro Truck Simulator 2

The route you drive along and how long it takes you along with traffic congestion is decided by the map you are on. A map of the Red Faction mars base with cars and shit would have a different layout for you to work with and different short cuts than europe.

>Borderlands

This is an FPS. Level Design in FPS is one of the most important things. Bad level design can make or break an FPS game. Your cover options, enemies, etc. are all decided by the map. The loot spawns, one of the most important things in the game, is also decided by the map.

>Every fighting game ever

Alright, you got me there. That's a good example of a few. However, SS13 isn't a closed up fighting game where the map is purely a background.

Shadowlight213 commented on code/controllers/subsystem/ticker.dm:
```diff
@@ -458,4 +461,79 @@ var/datum/subsystem/ticker/ticker
\ No newline at end of file
+			queue_delay = 0
+
+/datum/subsystem/ticker/proc/check_maprotate(var/force = 0)
```
The force argument for this proc looks like it got left over from the debug stuff, as its not used anywhere.

MrStonedOne replied:
Thank you fixing.

Mloc commented:
be careful with using shell()
it's known to lock up BYOND horribly

MrStonedOne commented:
@mloc It seems to be fine now, shell() is used by the bot to notify the admin channel ALOT.

Every time the shuttle docks at the station
Every time the round ends.
Every time somebody on the watch list joins when no admins are on
Every time an admin help comes in when no admins are on
Every time a first time player joins when no admins are on

MrStonedOne commented:
@Shadowlight213 not currently. But I'll look at adding a verb or some shit.

MrsTonedOne committed:
Fixed maprotate batch script having wrong exit codes

MrsTonedOne committed:
Removes unused force var from map rotate proc

Dorsisdwarf commented:
3 hours.

Mloc commented:
it might just be linux actually- and since this calls bat files you're already breaking platform compatibility :p

MrStonedOne commented:
you can run bat files in linux!

They would have to contain a shebang and bash code, but its doable =P

Strathcona commented:
Damn this is awesome! Will this make testing maps like dreamstation or whatever the mapping section spits out easier?

Razharas commented:
@MrStonedOne 
Tell when this is ready

Dorsisdwarf commented:
Live test when

MrsTonedOne committed:
Some map rotation tweaks.

Adds a config option for the map rotation chance.
Ups default rotation chance from 50% of round length to 75% of round length.
Moves the procs around.
Adds two new admin verbs:
Force map rotation.
Change map.

MrStonedOne commented:
>Live test when
Coming up. I'm gonna make another PR with all 3 maps to be tested, and this change as well. Then I'll close the pr and do a live test of it.

MrStonedOne commented:
#12111

Metacide commented:
Incredible! It works very well, and seamlessly too by the looks of things. I never really thought anyone'd get round to doing this. A bit more motivation to mappers, too. Thanks very much, really! :+1:

Strathcona commented:
I agree with Meta here, this is probably one of the best improvements to the core game in a looooong time.

freerealestate commented on code/world.dm:
```diff
@@ -294,3 +297,77 @@ var/failed_db_connections = 0
+	message_admins("Randomly rotating map to [VM.name]([VM.friendlyname])")
+	. = changemap(VM)
+	if (. == 0)
+		world << "<span class='boldannounce'>Map rotation has choosen [VM.friendlyname] for next round!</span>"
```
it's uh

'chosen'

freerealestate commented on code/modules/admin/verbs/maprotation.dm:
```diff
@@ -0,0 +1,34 @@
+			mapname += "\]"
+
+		maprotatechoices[mapname] = VM
+	var/choosenmap = input("Choose a map to rotate to", "Change Map")  as null|anything in maprotatechoices
```
chosen

Jordie0608 commented:
An aside suggestion: add an indicator to Show Server Revision if map rotation is enabled or not.

Razharas commented:
@MrStonedOne 
Is this ready?

MrStonedOne commented:
No, It has some bugs still. I'll have them mostly fixed. I'm trying to figure out how to fix the last one in a way that doesn't suck.

MrsTonedOne committed:
Map rotation fixes:
Fixes mutiple rotate operations from breaking the server, map rotation will properly detect rotation operations as well as update operations and abort.
Fixes max pop being treated as minpop
Adds more feedback to admin verbs.

MrsTonedOne committed:
adds prompt and logging to trigger map rotation verb

MrsTonedOne committed:
Adds a next map status tab entry.

MrsTonedOne committed:
Fixes typo in batch file. if exists -> if exist

MrsTonedOne committed:
Fixes added spam in server-tools's copyfromgit.bat

MrStonedOne commented:
This is as ready as it will ever be for merge.

MrsTonedOne committed:
choosen -> chosen

Giacom commented:
I'm not clear on how this works. I'm currently guessing it does this:

* Shuttle hits point of no return.
* Server starts compiling a new game with the new map.
* The current round ends.
* Server restarts and loads the new map.

If so, then is there a long downtime between rounds?

GunHog commented:
So far it has been pretty seamless.

Anonus commented:
IT IS DONE LOOK AT ALL THESE TWO MAPS IN ROTATION

Jordie0608 commented:
There are four.

freerealestate commented:
where's asteroid and uterus tho

Shadowlight213 commented:
Asteroid station is completely broken and was deleted from the repo.

Giacom commented:
https://github.com/tgstation/-tg-station/blob/master/config/maps.txt says 3 maps.

RemieRichards commented:
@Giacom it's a .txt config file, they're almost ALWAYS different on the live servers to the one on the repo.

freerealestate commented:
>Asteroid station is completely broken and was deleted from the repo.

RIP

What was broken about it, out of curiosity?

Shadowlight213 commented:
It ended up even failing to load in dreammaker, which meant it was impossible to continue work on.
